### ISSUE TITLE
Any ideas to run on 8 V100-32G gpus?

### ISSUE THREAD

**Issue by civat** [Deep-Agent/R1-V#89]
I tried:
1. offloading (optimizer state and parameters)
2. gradient checkpointing
3. reduce the max prompt length (256 now)
4. per_device_train_batch_size = 1
5. num_generations=1

OOM still occurs.
Any ideas to solve? (No A100 available now)

Thank you very much!

**Comment by Quinn777** [comment]
use lora

**Comment by m-Just** [comment]
Freezing the vision encoder saves a lot of memory.

**Comment by happyz123456789** [comment]
Could you teach me how do you solve  it?

**Comment by civat** [comment]
> Could you teach me how do you solve it?

Use PyTorch's torch.utils.checkpoint

Most parts of the full model do not support the gradient checkpointing even if gradient_checkpointing=True in the config file. So you need to add torch.utils.checkpoint in the expensive computation in the modeling_qwen2_vl.py file. 

After that, the training memory at the first step is about 15GB, which is a huge reduction.
But there is a new issue. I found that the memory increases after several training steps, and OOM occurs at the 16-th step. I do not why. I am trying to profile the memory usage.

**Comment by happyz123456789** [comment]
thank you very much

**Comment by happyz123456789** [comment]
> > Could you teach me how do you solve it?
> 
> Use PyTorch's torch.utils.checkpoint
> 
> Most parts of the full model do not support the gradient checkpointing even if gradient_checkpointing=True in the config file. So you need to add torch.utils.checkpoint in the expensive computation in the modeling_qwen2_vl.py file.
> 
> After that, the training memory at the first step is about 15GB, which is a huge reduction. But there is a new issue. I found that the memory increases after several training steps, and OOM occurs at the 16-th step. I do not why. I am trying to profile the memory usage.

thank you very much

**Comment by BIGBALLON** [comment]
@civat 

> But there is a new issue. I found that the memory increases after several training steps, and OOM occurs at the 16-th step. I do not why. I am trying to profile the memory usage.

is there anythin solution?

**Comment by civat** [comment]
> [@civat](https://github.com/civat)
> 
> > But there is a new issue. I found that the memory increases after several training steps, and OOM occurs at the 16-th step. I do not why. I am trying to profile the memory usage.
> 
> is there anythin solution?

All reasons:
1. Most parts of the full model do not support the gradient checkpointing even if gradient_checkpointing=True in the config file.
2. There are some images with large size, resulting in long image patch sequence;
3. The arg --max_pixels 401408 does not work;
4. Memory will not be released immediately.

Solution:
1. Most parts of the full model do not support the gradient checkpointing even if gradient_checkpointing=True in the config file.
Add torch.utils.checkpoint in the expensive computation in the modeling_qwen2_vl.py file. Here is a simple example:
By default, the vision part use VisionSdpaAttention. We can checkpoint the expensive computation:

```
        # before
        # attn_output = F.scaled_dot_product_attention(q, k, v, attention_mask, dropout_p=0.0)
        # attn_output = attn_output.transpose(0, 1)
        # attn_output = attn_output.reshape(seq_length, -1)
        # attn_output = self.proj(attn_output)
        
        # after
        attn_output = checkpoint(F.scaled_dot_product_attention, q, k, v, attention_mask, use_reentrant=True)
        attn_output = attn_output.transpose(0, 1)
        attn_output = attn_output.reshape(seq_length, -1)
        attn_output = checkpoint(self.proj, attn_output, use_reentrant=True)
```
       
For Qwen2VLSdpaAttention, change it to:

```
        # before
        # attn_output = torch.nn.functional.scaled_dot_product_attention(
                #     query_states,
                #     key_states,
                #     value_states,
                #     attn_mask=causal_mask,
                #     dropout_p=self.attention_dropout if self.training else 0.0,
                #     is_causal=is_causal,
                # )

        # after
        def scaled_dot_product_attention_forward(q, k, v, attn_mask, dropout_p, is_causal):
                    return torch.nn.functional.scaled_dot_product_attention(
                                q,
                                k,
                                v,
                                attn_mask=attn_mask,
                                dropout_p=dropout_p,
                                is_causal=is_causal,
                            )
        attn_output = checkpoint(
            scaled_dot_product_attention_forward,
            query_states,
            key_states,
            value_states,
            causal_mask,
            self.attention_dropout if self.training else 0.0,
            is_causal,
            use_reentrant=True
        )
```

2. There are some images with large size, resulting in long image patch sequence
See next "3. The arg --max_pixels 401408 does not work;"

3. The arg --max_pixels 401408 does not work;
By deafult, the model uses max_pixels =12845056,  a very large value when using 8-V100 32G. The default patch size is 14*14, so it supports max patch len is 12845056/ (14 * 14) = 65536. 

The config file reduce this value to 401408  / (14 * 14) = 2048. But no effect.

What's even stranger is that setting the default value in the Qwen2VLGRPOTrainer (grpo_trainer.py) also does not work. The solution is to add one line in the init method:
```
def __init__(
        self,
        model: Union[str, PreTrainedModel],
        reward_funcs: Union[RewardFunc, list[RewardFunc]],
        args: GRPOConfig = None,
        train_dataset: Optional[Union[Dataset, IterableDataset]] = None,
        eval_dataset: Optional[Union[Dataset, IterableDataset, dict[str, Union[Dataset, IterableDataset]]]] = None,
        processing_class: Optional[PreTrainedTokenizerBase] = None,
        reward_processing_classes: Optional[Union[PreTrainedTokenizerBase, list[PreTrainedTokenizerBase]]] = None,
        callbacks: Optional[list[TrainerCallback]] = None,
        optimizers: tuple[Optional[torch.optim.Optimizer], Optional[torch.optim.lr_scheduler.LambdaLR]] = (None, None),
        peft_config: Optional["PeftConfig"] = None,
        # max_pixels: Optional[int] = 12845056,
        max_pixels: Optional[int] = 100352,  # setting the default value does not work, why?
        min_pixels: Optional[int] = 3136,
        attn_implementation: str = "flash_attention_2",
    ):
        max_pixels = 100352  # it's ok now, max patch token len is 512
        ...
```

4. Memory will not be released immediately.

Call gc in the compute_loss function ():

```
def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=None):
        if return_outputs:
            raise ValueError("The GRPOTrainer does not support returning outputs")
        import gc
        gc.collect()
        torch.cuda.empty_cache()    
        ...
```

The peak memory is around 20GB now. I will try to reduce it more.

**Comment by zijian0312** [comment]
could you teach me  why two 2B parameter models require a significant amount of GPU memory
